src/mmds/src/lib.rs

```diff
@@ -69,6 +69,15 @@ pub fn parse_request(request_bytes: &[u8]) -> Response {
                 );
             }
 
+            let method = request.method();
+            if method != Method::Get {
+                return build_response(
+                    request.http_version(),
+                    StatusCode::NotImplemented,
```

Comment:
We should be using the status code 405 Method Not Allowed. In micro-http it is okay to return the status code `NotImplemented` for POST or other methods that are actually not implemented. In MMDS we actually do not support other methods besides GET, so the status should reflect this.

Reply:
If we return 405 then we must also include an `Allow` header field in the response with a list of the currently supported methods.

Reply:
Why complicate? I would return bad request, because this is what it is, a bad request.

Reply:
> The server MUST generate an Allow header field in a 405 response containing a list of the target resource's currently supported methods.

It's somehow more to this. `micro-http` supports only `GET`, `PATCH` and `PUT`. This would mean that MMDS would understand these three, but do not allow `PATCH` and `PUT`. So `Method not allowed` would be a good fit here. However, there are other HTTP methods which are not supported by `micro-http`, hence by MMDS as well, so to be 100% accurate, we would have to respond from within MMDS with `Not implemented` status for `DELETE`, `HEAD`, `POST`, `UPDATE`.

I think it is worth to make the distinction that this part here, where we create a response based on a request, is the actual HTTP server logic, which is reachable from within the guest, and we can choose our own response logic/content, depending on how the server perceives the request. It doesn't have to give insight into the support or the limitations of the `micro-http` crate, as far as I can tell, but I might be wrong.

I see several options here:
* choose `NotImplemented` status, which is a valid response status even if we understand some HTTP methods which are not allowed.
* choose `Method not allowed` status, which fits the best for this use case, and add support in `micro-http` as well.
* choose `Bad Request` status, which can bring the needed information to user, to understand that the method is not allowed. This can be seen as an client error, in that the request is deceptive, given the contract for MMDS guest applications requests.
* choose a combination of `NotImplement` and `Method not allowed`, with the considerations given at the beginning of the comment.

If am fine with all these options, but if it was my choice, I would simplify, and make it work with what we currently have, meaning `NotImplemented` or `Bad Request`. Please, let me know what you think.

Reply:
400 has different semantics: the api server cannot process the request due to a an error on the client side, for example malformed request syntax.

Reply:
I agree, it's not complicated. 

Other than this, where do we make the separation between HTTP methods which are not allowed, and functionality which is not yet implemented? Are these methods really not allowed? Why shouldn't be these methods allowed?

I am thinking that in the future there might come a need to support Firecracker MMDS user data. This would mean to support few HTTP methods like `PUT mmds_ipv4_addr/user-data/resource` and `PATCH mmds_ipv4_addr/user-data/resource`. However, should we simply assume in terms of allowing or not allowing these methods, or in terms of supporting functionality which would require these methods?

Please, let me know what you think.

Reply:
Synced offline with @sandreim . The scenario I've described can introduce many security holes, and it is not how IMDS works when providing user data at instance runtime. User-data has to be configured/modified when the instance is stopped, or before launching it. This means that the response `Method not allowed` status is necessary for other HTTP supported methods than `GET`.